File: backend/src/Domino.Backend/Application/Matches/MatchesService.cs
using Domino.Backend.Application.Matches.Models;
using Domino.Backend.Utilities;
using Microsoft.EntityFrameworkCore;

namespace Domino.Backend.Application.Matches;

public class MatchesService(DominoDbContext db) : IMatchesService
{
#pragma warning disable CA1845
    public async Task<List<MatchSummaryDto>> ListForUserAsync(int userId)
    {
        return await db.MatchUsers
            .Where(mu => mu.UserId == userId)
            .OrderByDescending(mu => mu.CreatedAt)
            .Select(mu => new MatchSummaryDto
            {
                PublicId = mu.Match!.PublicId,
                OtherUserName = db.MatchUsers
                    .Where(other => other.MatchId == mu.MatchId && other.UserId != userId)
                    .Select(other => other.User!.FirstName + " " + other.User!.LastName.Substring(0, 1) + ".")
                    .FirstOrDefault() ?? "Unknown",
                Status = mu.Match.CreatedAt.AddHours(24) < DateTime.UtcNow
                            && mu.Match.MatchUsers.Any(m => m.Accepted == null)
                    ? "expired"
                    : mu.Match.MatchUsers.All(m => m.Accepted == true)
                        ? "accepted"
                        : mu.Match.MatchUsers.Any(m => m.Accepted == false)
                            ? "denied"
                            : "pending",
                CreatedAt = mu.CreatedAt
            })
            .ToListAsync();
    }
#pragma warning restore CA1845

    public async Task<ServiceResult<string>> CreateAsync(CreateMatchRequest request, int createdByUserId)
    {
        if (request.UserId1 == request.UserId2)
        {
            return ServiceResult.Invalid<string>("Cannot match a user with themselves.");
        }

        if (request.ActivityIdeaIds.Count != 3)
        {
            return ServiceResult.Invalid<string>("Exactly 3 activity ideas are required.");
        }

        if (request.ActivityIdeaIds.Distinct().Count() != 3)
        {
            return ServiceResult.Invalid<string>("Activity ideas must be unique.");
        }

        var user1Exists = await db.Users.AnyAsync(u => u.Id == request.UserId1 && u.IsActive);
        var user2Exists = await db.Users.AnyAsync(u => u.Id == request.UserId2 && u.IsActive);
        if (!user1Exists || !user2Exists)
        {
            return ServiceResult.Invalid<string>("One or both users not found.");
        }

        var validIdeaCount = await db.ActivityIdeas
            .CountAsync(a => request.ActivityIdeaIds.Contains(a.Id));
        if (validIdeaCount != 3)
        {
            return ServiceResult.Invalid<string>("One or more activity ideas not found.");
        }

        var match = new MatchModel
        {
            Narrative = request.Narrative,
            CreatedByUserId = createdByUserId
        };

        db.Matches.Add(match);
        await db.SaveChangesAsync();

        db.MatchUsers.AddRange(
            new MatchUserModel { MatchId = match.Id, UserId = request.UserId1 },
            new MatchUserModel { MatchId = match.Id, UserId = request.UserId2 }
        );

        db.MatchActivityIdeas.AddRange(
            request.ActivityIdeaIds.Select(ideaId => new MatchActivityIdeaModel
            {
                MatchId = match.Id,
                ActivityIdeaId = ideaId
            })
        );

        await db.SaveChangesAsync();

        return ServiceResult.Success(match.PublicId);
    }

    public async Task<ServiceResult<MatchDetailDto>> GetAsync(string publicId, int currentUserId, bool isAdmin)
    {
        var match = await db.Matches
            .Include(m => m.MatchUsers).ThenInclude(mu => mu.User)
            .Include(m => m.MatchActivityIdeas).ThenInclude(mai => mai.ActivityIdea)
            .FirstOrDefaultAsync(m => m.PublicId == publicId);

        if (match is null)
        {
            return ServiceResult.NotFound<MatchDetailDto>("Match not found.");
        }

        var isParticipant = match.MatchUsers.Any(mu => mu.UserId == currentUserId);
        if (!isParticipant && !isAdmin)
        {
            return ServiceResult.NotFound<MatchDetailDto>("Match not found.");
        }

        var isExpired = DateTime.UtcNow > match.CreatedAt.AddHours(24);
        var bothAccepted = match.MatchUsers.All(mu => mu.Accepted == true);

        var users = match.MatchUsers.Select(mu => new MatchUserDto
        {
            UserId = mu.UserId,
            FirstName = mu.User!.FirstName,
            LastInitial = mu.User.LastName.Length > 0
                ? string.Concat(mu.User.LastName.AsSpan(0, 1), ".")
                : "",
            Age = CalculateAge(mu.User.Birthday),
            Accepted = mu.Accepted
        }).ToList();

        var dto = new MatchDetailDto
        {
            PublicId = match.PublicId,
            Narrative = match.Narrative,
            Users = users,
            IsExpired = isExpired,
            BothAccepted = bothAccepted,
            ActivityIdeas = bothAccepted
                ? match.MatchActivityIdeas.Select(mai => new MatchActivityIdeaDto
                {
                    Id = mai.ActivityIdea!.Id,
                    Name = mai.ActivityIdea.Name,
                    Description = mai.ActivityIdea.Description
                }).ToList()
                : [],
            CreatedAt = match.CreatedAt,
            CurrentUserAccepted = match.MatchUsers
                .FirstOrDefault(mu => mu.UserId == currentUserId)?.Accepted
        };

        return ServiceResult.Success(dto);
    }

    public async Task<ServiceResult<RespondResult>> RespondAsync(string publicId, int currentUserId, bool accepted)
    {
        var match = await db.Matches
            .Include(m => m.MatchUsers)
            .FirstOrDefaultAsync(m => m.PublicId == publicId);

        if (match is null)
        {
            return ServiceResult.NotFound<RespondResult>("Match not found.");
        }

        var matchUser = match.MatchUsers.FirstOrDefault(mu => mu.UserId == currentUserId);
        if (matchUser is null)
        {
            return ServiceResult.NotFound<RespondResult>("Match not found.");
        }

        if (DateTime.UtcNow > match.CreatedAt.AddHours(24))
        {
            return ServiceResult.Invalid<RespondResult>(
                "This match has expired. Responses are no longer accepted.");
        }

        if (matchUser.Accepted is not null)
        {
            return ServiceResult.Invalid<RespondResult>(
                "You have already responded to this match.");
        }

        matchUser.Accepted = accepted;
        await db.SaveChangesAsync();

        var bothAccepted = match.MatchUsers.All(mu => mu.Accepted == true);

        return ServiceResult.Success(
            new RespondResult { Accepted = accepted, BothAccepted = bothAccepted });
    }

    private static int CalculateAge(DateOnly birthday)
    {
        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        var age = today.Year - birthday.Year;
        if (birthday.AddYears(age) > today)
        {
            age--;
        }
        return age;
    }
}

File: backend-ts/src/matches/matches.service.ts
import { Injectable } from '@nestjs/common';
import { randomBytes } from 'crypto';
import { PrismaService } from '../prisma/prisma.service';
import type { ServiceResult } from '../common/service-result';
import { success, notFound, invalid } from '../common/service-result';
import type {
  MatchSummaryDto,
  MatchDetailDto,
  RespondResult,
  CreateMatchRequest,
} from './dto/match.dto';

const MATCH_EXPIRY_HOURS = 24;
const PUBLIC_ID_CHARS =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

function generatePublicId(length = 10): string {
  const bytes = randomBytes(length);
  return Array.from(bytes)
    .map((b) => PUBLIC_ID_CHARS[b % PUBLIC_ID_CHARS.length])
    .join('');
}

function calculateAge(birthday: Date): number {
  const today = new Date();
  let age = today.getFullYear() - birthday.getFullYear();
  const monthDiff = today.getMonth() - birthday.getMonth();
  if (
    monthDiff < 0 ||
    (monthDiff === 0 && today.getDate() < birthday.getDate())
  ) {
    age--;
  }
  return age;
}

function computeMatchStatus(
  createdAt: Date,
  matchUsers: { accepted: boolean | null }[],
): string {
  const isExpired =
    new Date().getTime() - createdAt.getTime() > MATCH_EXPIRY_HOURS * 3600000;

  if (isExpired && matchUsers.some((mu) => mu.accepted === null)) {
    return 'expired';
  }
  if (matchUsers.every((mu) => mu.accepted === true)) return 'accepted';
  if (matchUsers.some((mu) => mu.accepted === false)) return 'denied';
  return 'pending';
}

@Injectable()
export class MatchesService {
  constructor(private readonly prisma: PrismaService) {}

  async listForUser(userId: number): Promise<MatchSummaryDto[]> {
    const matchUsers = await this.prisma.matchUser.findMany({
      where: { userId },
      orderBy: { createdAt: 'desc' },
      include: {
        match: {
          include: {
            matchUsers: {
              include: { user: true },
            },
          },
        },
      },
    });

    return matchUsers.map((mu) => {
      const otherUser = mu.match.matchUsers.find(
        (m) => m.userId !== userId,
      )?.user;
      const otherName = otherUser
        ? `${otherUser.firstName} ${otherUser.lastName.substring(0, 1)}.`
        : 'Unknown';

      return {
        publicId: mu.match.publicId,
        otherUserName: otherName,
        status: computeMatchStatus(mu.match.createdAt, mu.match.matchUsers),
        createdAt: mu.createdAt,
      };
    });
  }

  async create(
    request: CreateMatchRequest,
    createdByUserId: number,
  ): Promise<ServiceResult<string>> {
    if (request.userId1 === request.userId2) {
      return invalid('Cannot match a user with themselves.');
    }

    if (new Set(request.activityIdeaIds).size !== 3) {
      return invalid('Activity ideas must be unique.');
    }

    const [user1, user2] = await Promise.all([
      this.prisma.user.findFirst({
        where: { id: request.userId1, isActive: true },
      }),
      this.prisma.user.findFirst({
        where: { id: request.userId2, isActive: true },
      }),
    ]);

    if (!user1 || !user2) {
      return invalid('One or both users not found.');
    }

    const validIdeaCount = await this.prisma.activityIdea.count({
      where: { id: { in: request.activityIdeaIds } },
    });

    if (validIdeaCount !== 3) {
      return invalid('One or more activity ideas not found.');
    }

    const publicId = generatePublicId();

    const match = await this.prisma.match.create({
      data: {
        publicId,
        narrative: request.narrative,
        createdByUserId,
      },
    });

    await this.prisma.matchUser.createMany({
      data: [
        { matchId: match.id, userId: request.userId1 },
        { matchId: match.id, userId: request.userId2 },
      ],
    });

    await this.prisma.matchActivityIdea.createMany({
      data: request.activityIdeaIds.map((ideaId) => ({
        matchId: match.id,
        activityIdeaId: ideaId,
      })),
    });

    return success(publicId);
  }

  async get(
    publicId: string,
    currentUserId: number,
    isAdmin: boolean,
  ): Promise<ServiceResult<MatchDetailDto>> {
    const match = await this.prisma.match.findUnique({
      where: { publicId },
      include: {
        matchUsers: { include: { user: true } },
        matchActivityIdeas: { include: { activityIdea: true } },
      },
    });

    if (!match) {
      return notFound('Match not found.');
    }

    const isParticipant = match.matchUsers.some(
      (mu) => mu.userId === currentUserId,
    );
    if (!isParticipant && !isAdmin) {
      return notFound('Match not found.');
    }

    const isExpired =
      new Date().getTime() - match.createdAt.getTime() >
      MATCH_EXPIRY_HOURS * 3600000;
    const bothAccepted = match.matchUsers.every(
      (mu) => mu.accepted === true,
    );

    const users = match.matchUsers.map((mu) => ({
      userId: mu.userId,
      firstName: mu.user.firstName,
      lastInitial:
        mu.user.lastName.length > 0
          ? `${mu.user.lastName.substring(0, 1)}.`
          : '',
      age: calculateAge(mu.user.birthday),
      accepted: mu.accepted,
    }));

    return success({
      publicId: match.publicId,
      narrative: match.narrative,
      users,
      isExpired,
      bothAccepted,
      activityIdeas: bothAccepted
        ? match.matchActivityIdeas.map((mai) => ({
            id: mai.activityIdea.id,
            name: mai.activityIdea.name,
            description: mai.activityIdea.description,
          }))
        : [],
      createdAt: match.createdAt,
      currentUserAccepted:
        match.matchUsers.find((mu) => mu.userId === currentUserId)
          ?.accepted ?? null,
    });
  }

  async respond(
    publicId: string,
    currentUserId: number,
    accepted: boolean,
  ): Promise<ServiceResult<RespondResult>> {
    const match = await this.prisma.match.findUnique({
      where: { publicId },
      include: { matchUsers: true },
    });

    if (!match) {
      return notFound('Match not found.');
    }

    const matchUser = match.matchUsers.find(
      (mu) => mu.userId === currentUserId,
    );
    if (!matchUser) {
      return notFound('Match not found.');
    }

    const isExpired =
      new Date().getTime() - match.createdAt.getTime() >
      MATCH_EXPIRY_HOURS * 3600000;
    if (isExpired) {
      return invalid(
        'This match has expired. Responses are no longer accepted.',
      );
    }

    if (matchUser.accepted !== null) {
      return invalid('You have already responded to this match.');
    }

    await this.prisma.matchUser.update({
      where: { id: matchUser.id },
      data: { accepted },
    });

    // Re-check all match users for bothAccepted
    const updatedMatchUsers = match.matchUsers.map((mu) =>
      mu.id === matchUser.id ? { ...mu, accepted } : mu,
    );
    const bothAccepted = updatedMatchUsers.every(
      (mu) => mu.accepted === true,
    );

    return success({ accepted, bothAccepted });
  }
}
